File: apps/server/src/processes/sampler.ts
import { errorText } from '../error-text.ts';
/*
 * One reading of the process table. Every counter is cumulative, so a rate is always the difference
 * between two readings over the time between them; that is what keeps a five minute gap honest.
 */

export interface RawProcess {
    pid: number;
    ppid: number;
    uid: number;
    /* Microseconds since the epoch. */
    startTime: number;
    name: string;
    path: string | null;
    /* False when the kernel refused the counters (another user, the system); they are null then. */
    readable: boolean;
    cpuNs: number | null;
    memory: number | null;
    diskRead: number | null;
    diskWrite: number | null;
}

export interface MachineCounters {
    cores: number;
    /* Cumulative ticks over all cores; a share needs two readings. */
    cpuBusy: number | null;
    cpuTotal: number | null;
    memoryUsed: number | null;
    memoryTotal: number;
    diskFree: number | null;
    diskTotal: number | null;
}

export interface RawSample {
    /* Epoch milliseconds. */
    at: number;
    /* A clock that stops while the machine sleeps; every elapsed time is measured on it. */
    awakeMs: number;
    /* The boot clock minus the awake clock: it only grows while the machine sleeps. */
    asleepMs: number;
    processes: RawProcess[];
    machine: MachineCounters;
}

export interface CommandLine {
    args: string[];
    env: Record<string, string>;
}

export interface ProcessSampler {
    sample(): RawSample;
    /* One process, for the check right before a signal. */
    inspect(pid: number): { startTime: number; uid: number } | null;
    /* The arguments and environment of a process of this user; null when the kernel refuses. */
    commandLine(pid: number): CommandLine | null;
}

export interface ProcessRate {
    /* Percent of one core. */
    cpu: number | null;
    memory: number | null;
    /* Bytes per second. */
    diskRead: number | null;
    diskWrite: number | null;
    /* The process was not in the reading before; a new process is a sign of life. */
    spawned: boolean;
}

export interface MachineRate {
    /* Percent of all cores. */
    cpu: number | null;
}

/* More sleep than this between two readings and the history starts over; less is clock jitter. */
export const SLEEP_TOLERANCE_MS = 2000;

export const identityOf = (pid: number, startTime: number): string => `${pid}:${startTime}`;

/* Mach ticks to nanoseconds. On Apple silicon a tick is 125/3 ns, so the raw number reads 40 times too low. */
export const ticksToNs = (ticks: number, numer: number, denom: number): number => (ticks * numer) / denom;

/* Whether the machine slept between two readings. How long the gap was says nothing: a closed panel samples every five minutes. */
export const sleptBetween = (before: RawSample, after: RawSample): boolean => after.asleepMs - before.asleepMs > SLEEP_TOLERANCE_MS;

const perSecond = (now: number | null, then: number | null, elapsedMs: number): number | null =>
    now === null || then === null ? null : (Math.max(0, now - then) * 1000) / elapsedMs;

/*
 * The rate of every process over the time since `before`. A process that started after `before` was
 * taken spent all its counters inside the interval, so it has a rate on its first sight too; one
 * that was already running but not in `before` has none yet, because nothing says when it spent them.
 */
export const processRates = (before: RawSample | null, after: RawSample): Map<string, ProcessRate> => {
    const rates = new Map<string, ProcessRate>();
    const previous = new Map<string, RawProcess>();
    for (const process of before?.processes ?? []) {
        previous.set(identityOf(process.pid, process.startTime), process);
    }
    const elapsedMs = before === null ? 0 : after.awakeMs - before.awakeMs;
    for (const process of after.processes) {
        const identity = identityOf(process.pid, process.startTime);
        const seen = previous.get(identity);
        if (!process.readable) {
            rates.set(identity, { cpu: null, memory: null, diskRead: null, diskWrite: null, spawned: false });
            continue;
        }
        const bornInside = before !== null && seen === undefined && process.startTime >= before.at * 1000;
        const base = seen?.readable ? seen : bornInside ? { cpuNs: 0, diskRead: 0, diskWrite: 0 } : null;
        const measurable = base !== null && elapsedMs > 0;
        rates.set(identity, {
            cpu: measurable && process.cpuNs !== null && base.cpuNs !== null ? (Math.max(0, process.cpuNs - base.cpuNs) / 1e6 / elapsedMs) * 100 : null,
            memory: process.memory,
            diskRead: measurable ? perSecond(process.diskRead, base.diskRead, elapsedMs) : null,
            diskWrite: measurable ? perSecond(process.diskWrite, base.diskWrite, elapsedMs) : null,
            spawned: before !== null && seen === undefined
        });
    }
    return rates;
};

export const machineRate = (before: RawSample | null, after: RawSample): MachineRate => {
    const then = before?.machine;
    const now = after.machine;
    if (!then || then.cpuBusy === null || then.cpuTotal === null || now.cpuBusy === null || now.cpuTotal === null || now.cpuTotal <= then.cpuTotal) {
        return { cpu: null };
    }
    return { cpu: (Math.max(0, now.cpuBusy - then.cpuBusy) / (now.cpuTotal - then.cpuTotal)) * 100 };
};

/*
 * `KERN_PROCARGS2`: the argument count, the executable path and its padding, the arguments, then the
 * environment, every string closed by a zero byte.
 */
export const parseProcArgs = (bytes: Uint8Array): CommandLine | null => {
    if (bytes.byteLength < 4) {
        return null;
    }
    const argc = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength).getInt32(0, true);
    let offset = 4;
    while (offset < bytes.byteLength && bytes[offset] !== 0) {
        offset++;
    }
    while (offset < bytes.byteLength && bytes[offset] === 0) {
        offset++;
    }
    const decoder = new TextDecoder();
    const strings: string[] = [];
    let start = offset;
    for (let i = offset; i < bytes.byteLength; i++) {
        if (bytes[i] !== 0) {
            continue;
        }
        if (i === start && strings.length >= argc) {
            // A double zero after the environment is where the area ends.
            break;
        }
        strings.push(decoder.decode(bytes.subarray(start, i)));
        start = i + 1;
    }
    return { args: strings.slice(0, argc), env: parseEnvironment(strings.slice(argc)) };
};

export const parseEnvironment = (entries: readonly string[]): Record<string, string> => {
    const env: Record<string, string> = {};
    for (const entry of entries) {
        const equals = entry.indexOf('=');
        if (equals > 0) {
            env[entry.slice(0, equals)] = entry.slice(equals + 1);
        }
    }
    return env;
};

/* The sampler of this platform, or null where there is none (Windows); the panel explains that instead. */
export const createSampler = async (platform: string = process.platform, home: string = process.env.HOME ?? '/'): Promise<ProcessSampler | null> => {
    try {
        if (platform === 'darwin') {
            const { DarwinSampler } = await import('./darwin.ts');
            return new DarwinSampler(home);
        }
        if (platform === 'linux') {
            const { LinuxSampler } = await import('./linux.ts');
            return new LinuxSampler(home);
        }
    } catch (e) {
        console.error('The process sampler could not start:', errorText(e));
    }
    return null;
};

File: crates/ruimte_runtime/src/processes/sampler.rs
use std::{collections::HashMap, path::Path, sync::Arc};

#[derive(Clone, Debug, Default)]
pub struct CommandLine {
    pub args: Vec<String>,
    pub env: HashMap<String, String>,
}

#[derive(Clone, Debug)]
pub struct RawProcess {
    pub pid: u32,
    pub ppid: u32,
    pub uid: u32,
    pub start_time: u64,
    pub name: String,
    pub path: Option<String>,
    pub readable: bool,
    pub cpu_ns: Option<u64>,
    pub memory: Option<u64>,
    pub disk_read: Option<u64>,
    pub disk_write: Option<u64>,
}

#[derive(Clone, Debug)]
pub struct MachineCounters {
    pub cores: usize,
    pub cpu_busy: Option<u64>,
    pub cpu_total: Option<u64>,
    pub memory_used: Option<u64>,
    pub memory_total: u64,
    pub disk_free: Option<u64>,
    pub disk_total: Option<u64>,
}

#[derive(Clone, Debug)]
pub struct RawSample {
    pub at: u64,
    pub awake_ms: u64,
    pub asleep_ms: u64,
    pub processes: Vec<RawProcess>,
    pub machine: MachineCounters,
}

pub trait ProcessSampler: Send + Sync {
    fn sample(&self) -> anyhow::Result<RawSample>;
    fn inspect(&self, pid: u32) -> anyhow::Result<Option<(u64, u32)>>;
    fn command_line(&self, pid: u32) -> Option<CommandLine>;
}

pub fn parse_nul_command_line(bytes: &[u8], argc: Option<usize>) -> CommandLine {
    let values = bytes
        .split(|byte| *byte == 0)
        .filter(|value| !value.is_empty())
        .map(|value| String::from_utf8_lossy(value).into_owned())
        .collect::<Vec<_>>();
    let argc = argc.unwrap_or(values.len()).min(values.len());
    let args = values[..argc].to_vec();
    let env = values[argc..]
        .iter()
        .filter_map(|entry| entry.split_once('='))
        .filter(|(key, _)| !key.is_empty())
        .map(|(key, value)| (key.to_owned(), value.to_owned()))
        .collect();
    CommandLine { args, env }
}

#[derive(Clone, Debug)]
pub struct ProcessRate {
    pub cpu: Option<f64>,
    pub memory: Option<u64>,
    pub disk_read: Option<f64>,
    pub disk_write: Option<f64>,
}

pub fn create(home: &Path) -> Option<Arc<dyn ProcessSampler>> {
    #[cfg(target_os = "macos")]
    {
        return Some(Arc::new(super::darwin::DarwinSampler::new(home.to_owned())));
    }
    #[cfg(target_os = "linux")]
    {
        return Some(Arc::new(super::linux::LinuxSampler::new(home.to_owned())));
    }
    #[allow(unreachable_code)]
    None
}

pub fn identity(process: &RawProcess) -> (u32, u64) {
    (process.pid, process.start_time)
}

pub fn rates(
    before: Option<&RawSample>,
    after: &RawSample,
) -> std::collections::HashMap<(u32, u64), ProcessRate> {
    let previous: std::collections::HashMap<(u32, u64), &RawProcess> = before
        .map(|sample| {
            sample
                .processes
                .iter()
                .map(|process| (identity(process), process))
                .collect()
        })
        .unwrap_or_default();
    let elapsed_ms = before
        .map(|sample| after.awake_ms.saturating_sub(sample.awake_ms))
        .unwrap_or(0);
    let mut rates = std::collections::HashMap::new();
    for process in &after.processes {
        let seen = previous.get(&identity(process)).copied();
        let born_inside =
            before.is_some_and(|sample| seen.is_none() && process.start_time >= sample.at * 1000);
        let base = if seen.is_some_and(|process| process.readable) {
            seen
        } else {
            None
        };
        let measurable = elapsed_ms > 0 && (base.is_some() || born_inside);
        let base_cpu = base.and_then(|process| process.cpu_ns).unwrap_or(0);
        let base_read = base.and_then(|process| process.disk_read).unwrap_or(0);
        let base_write = base.and_then(|process| process.disk_write).unwrap_or(0);
        rates.insert(
            identity(process),
            ProcessRate {
                cpu: (process.readable && measurable)
                    .then(|| {
                        process.cpu_ns.map(|now| {
                            now.saturating_sub(base_cpu) as f64 / 1_000_000. / elapsed_ms as f64
                                * 100.
                        })
                    })
                    .flatten(),
                memory: process.readable.then_some(process.memory).flatten(),
                disk_read: (process.readable && measurable)
                    .then(|| {
                        process.disk_read.map(|now| {
                            now.saturating_sub(base_read) as f64 * 1000. / elapsed_ms as f64
                        })
                    })
                    .flatten(),
                disk_write: (process.readable && measurable)
                    .then(|| {
                        process.disk_write.map(|now| {
                            now.saturating_sub(base_write) as f64 * 1000. / elapsed_ms as f64
                        })
                    })
                    .flatten(),
            },
        );
    }
    rates
}

pub fn machine_cpu(before: Option<&RawSample>, after: &RawSample) -> Option<f64> {
    let before = before?;
    let busy = after
        .machine
        .cpu_busy?
        .saturating_sub(before.machine.cpu_busy?);
    let total = after
        .machine
        .cpu_total?
        .saturating_sub(before.machine.cpu_total?);
    (total > 0).then_some(busy as f64 / total as f64 * 100.)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn process(start_time: u64, cpu_ns: u64) -> RawProcess {
        RawProcess {
            pid: 7,
            ppid: 1,
            uid: 1,
            start_time,
            name: "test".into(),
            path: None,
            readable: true,
            cpu_ns: Some(cpu_ns),
            memory: Some(12),
            disk_read: Some(0),
            disk_write: Some(0),
        }
    }

    fn sample(at: u64, awake_ms: u64, processes: Vec<RawProcess>) -> RawSample {
        RawSample {
            at,
            awake_ms,
            asleep_ms: 0,
            processes,
            machine: MachineCounters {
                cores: 4,
                cpu_busy: None,
                cpu_total: None,
                memory_used: None,
                memory_total: 1,
                disk_free: None,
                disk_total: None,
            },
        }
    }

    #[test]
    fn pid_reuse_does_not_mix_counters() {
        let before = sample(1000, 1000, vec![process(900_000, 500_000_000)]);
        let after = sample(2000, 2000, vec![process(1_500_000, 100_000_000)]);
        let rate = rates(Some(&before), &after)
            .remove(&(7, 1_500_000))
            .unwrap();
        assert_eq!(rate.cpu, Some(10.0));
    }

    #[test]
    fn nul_command_line_separates_arguments_and_environment() {
        let line = parse_nul_command_line(
            b"node\0agent.js\0RUIMTE_SESSION_ID=session\0EMPTY=\0",
            Some(2),
        );

        assert_eq!(line.args, ["node", "agent.js"]);
        assert_eq!(line.env["RUIMTE_SESSION_ID"], "session");
        assert_eq!(line.env["EMPTY"], "");
    }
}
